File: docs/ezplm_api_manual/ezplm查询接口操作手册/demo/api-key-signing-demo.js
#!/usr/bin/env node
import crypto from 'node:crypto';

const apiKey = '';
const baseUrl = 'https://www.ezplm.cn';
const keyword = 'TPS79301DBVR';
const pageSize = '10';

function canonicalQuery(params) {
  return Object.entries(params)
    .filter(([, value]) => value !== undefined && value !== null && String(value) !== '')
    .map(([key, value]) => [String(key), String(value)])
    .sort(([leftKey, leftValue], [rightKey, rightValue]) =>
      leftKey === rightKey
        ? leftValue.localeCompare(rightValue)
        : leftKey.localeCompare(rightKey),
    )
    .map(
      ([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`,
    )
    .join('&');
}

function buildSignature({ method, path, params, timestamp, nonce }) {
  const canonical = [
    method.toUpperCase(),
    path,
    canonicalQuery(params),
    timestamp,
    nonce,
  ].join('\n');
  return crypto
    .createHmac('sha256', apiKey)
    .update(canonical)
    .digest('base64url');
}

async function requestJson(path, params) {
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const nonce = crypto.randomUUID();
  const signature = buildSignature({ method: 'GET', path, params, timestamp, nonce });
  const query = canonicalQuery(params);
  const url = query ? `${baseUrl}${path}?${query}` : `${baseUrl}${path}`;

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'X-API-Key': apiKey,
      'X-Timestamp': timestamp,
      'X-Nonce': nonce,
      'X-Signature': signature,
    },
  });

  const body = await response.text();
  let parsed;
  try {
    parsed = JSON.parse(body);
  } catch {
    parsed = { raw: body };
  }
  return { status: response.status, body: parsed };
}

const parts = await requestJson('/api/v1/api-key/parts', {
  keyword,
  pageSize,
});
console.log('parts_status =', parts.status);
console.log(JSON.stringify(parts.body, null, 2));

if (!Array.isArray(parts.body?.data) || parts.body.data.length === 0) {
  console.error('no part returned');
  process.exit(2);
}

const partlibId = parts.body.data[0].id;
const refs = await requestJson('/api/v1/api-key/reference-designs', {
  partlibId,
  pageSize,
});
console.log('reference_designs_status =', refs.status);
console.log(JSON.stringify(refs.body, null, 2));

process.exit(parts.status === 200 && refs.status === 200 ? 0 : 3);

File: docs/ezplm_api_manual/ezplm查询接口操作手册/demo/api-key-signing-demo.py
#!/usr/bin/env python3
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import secrets
import sys
import time
import uuid
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, Iterable, Tuple

API_KEY = ""
BASE_URL = "https://www.ezplm.cn"
PART_KEYWORD = "TPS79301DBVR"
PAGE_SIZE = "10"


def canonical_query(params: Dict[str, Any]) -> str:
    items: list[Tuple[str, str]] = []
    for key, value in params.items():
        if value is None or value == "":
            continue
        items.append((str(key), str(value)))
    items.sort(key=lambda item: (item[0], item[1]))
    return "&".join(
        f"{urllib.parse.quote(key, safe='')}={urllib.parse.quote(value, safe='')}"
        for key, value in items
    )


def build_signature(
    api_key: str,
    method: str,
    path: str,
    params: Dict[str, Any],
    timestamp: str,
    nonce: str,
) -> str:
    canonical = "\n".join(
        [method.upper(), path, canonical_query(params), timestamp, nonce]
    )
    digest = hmac.new(
        api_key.encode("utf-8"),
        canonical.encode("utf-8"),
        hashlib.sha256,
    ).digest()
    return base64.urlsafe_b64encode(digest).decode("utf-8").rstrip("=")


def request_json(
    base_url: str,
    api_key: str,
    path: str,
    params: Dict[str, Any],
) -> tuple[int, dict[str, Any]]:
    timestamp = str(int(time.time()))
    nonce = str(uuid.uuid4())
    signature = build_signature(api_key, "GET", path, params, timestamp, nonce)
    query = canonical_query(params)
    url = f"{base_url}{path}"
    if query:
        url = f"{url}?{query}"

    req = urllib.request.Request(
        url,
        method="GET",
        headers={
            "X-API-Key": api_key,
            "X-Timestamp": timestamp,
            "X-Nonce": nonce,
            "X-Signature": signature,
        },
    )

    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            body = resp.read().decode("utf-8")
            return resp.status, json.loads(body)
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8")
        try:
            parsed = json.loads(body)
        except json.JSONDecodeError:
            parsed = {"raw": body}
        return exc.code, parsed


def main() -> int:
    api_key = API_KEY.strip()
    base_url = BASE_URL.rstrip("/")
    keyword = PART_KEYWORD
    page_size = PAGE_SIZE

    parts_status, parts_body = request_json(
        base_url,
        api_key,
        "/api/v1/api-key/parts",
        {"keyword": keyword, "pageSize": page_size},
    )
    print("parts_status =", parts_status)
    print(json.dumps(parts_body, ensure_ascii=False, indent=2))

    items = parts_body.get("data", [])
    if not items:
        print("no part returned", file=sys.stderr)
        return 2

    partlib_id = items[0]["id"]
    ref_status, ref_body = request_json(
        base_url,
        api_key,
        "/api/v1/api-key/reference-designs",
        {"partlibId": partlib_id, "pageSize": page_size},
    )
    print("reference_designs_status =", ref_status)
    print(json.dumps(ref_body, ensure_ascii=False, indent=2))
    return 0 if parts_status == 200 and ref_status == 200 else 3


if __name__ == "__main__":
    raise SystemExit(main())
